File: assign1/hashtable.py
#@author Adam Voliva
#@description CISP 430 Data Structures
#@name Assignment 1 Hash Table
#@date January 31, 2013


import pickle

MaxSlots = 3
MaxBuckets = 30
TableSize = 20
OverflowSize = 10


class Slot(dict):

    def __init__(self):
        self.key = None
        self.value = None


class Bucket:

    def __init__(self):
        self.count = 0
        self.overflow = None
        self.slots = []
        for i in range(0, 3):
            self.slots.append({})


class HashTable:

    def __init__(self):
        self.table = []
        for i in range(0, MaxBuckets):
            self.table.append(Bucket())

    def Search(self):
        f = open('./search.dat', 'r+')
        f2 = open('./retrieval.txt', 'w+')
        f2.write('Search and Retrieval'.center(82))
        f2.write('\n')
        f2.write('Transactions'.center(82))
        f2.write('\n')
        f2.write('Search Key\t\tBucket/Slot\t\tRecord'.center(82))
        f2.write('\n')
        for keys in f.readlines():
            key = self.Hash(keys[:10])
            for pos, item in enumerate(self.table):
                for ps, ite in enumerate(item.slots):
                    if key in item.slots[ps]:
                        if keys[:10] == item.slots[ps][key].key:
                            f2.write('{0:>35}{1:>16}/{2}\t\t{3}'.format(
                                item.slots[ps][key].key,
                                pos + 1,
                                ps + 1,
                                item.slots[ps][key].value
                            ))
                            f2.write('\n')
        f.close()
        f2.close()

    def RestoreReport(self):
        if self.table is not None:
            raise Exception('The hash table has been restored already.')
        f = open('./hashtable.txt', 'r+')
        self.table = pickle.load(f)
        f.close()

    def SaveReport(self):
        f = open('./hashtable.txt', 'w+')
        # Pickle dumps a data structure to a file
        pickle.dump(self.table, f)
        f.close()
        # Deleted from memory
        self.table = None

    def GenerateReport(self):
        f = open('./before.txt', 'w+')
        f.write('Hash Table'.center(82))
        f.write('\n')
        f.write('Verififcation Report'.center(82))
        f.write('\n')
        f.write('Before Restoration'.center(82))
        f.write('\n')
        for pos, item in enumerate(self.table):
            f.write('Bucket {0}\n'.format(pos + 1))
            for ps, ite in enumerate(item.slots):
                for key, value in item.slots[ps].iteritems():
                    if item.slots[ps][key] is not None:
                        f.write('\tSlot {0}: {1}{2}\n'.format(ps + 1,
                                item.slots[ps][key].key,
                                item.slots[ps][key].value))
                if len(item.slots[ps].keys()) is 0:
                    f.write('\tSlot {0}: \n'.format(ps + 1))
            if self.table[pos].overflow is not None:
                f.write('Overflow Pointer: {0}\n'.format(
                    int(self.table[pos].overflow) + 1))
            else:
                f.write('Overflow Pointer: {0}\n'.format(
                        self.table[pos].overflow))
            f.write('\n')

    def DataIn(self):
        f = open('./datain.dat', 'r+')
        for line in f.readlines():
            self.Insert(self.Hash(line[:10]), line[10:30], line[:10])
        f.close()

    def Hash(self, key):
        return (ord(key[2]) + ord(key[4]) + ord(key[6])) % TableSize

    def Insert(self, key, value, data):
        tup = self.CollissionCheck(key)
        if tup is not None:
            # collission is true
            item, ht_index, slot_index = tup
            # index = None
            # print slot_index
            # print self.table[ht_index].count
            # print item
            if self.table[ht_index].count > 2:
                ht_index, slot_index = self.CheckAvailableBuckets(key, 19, MaxBuckets)
                # overflow is true
                # length = self.GetTableLength(key, index)
                length = self.table[ht_index].count
                while length == 3:
                    # overflow of overflow is true
                    ht_index, slot_index = self.CheckAvailableBuckets(key,
                        ht_index + 1, MaxBuckets)
                    length = self.table[ht_index].count
                    # length = self.GetTableLength(key, index)
                if length > 0:
                    # insert into 'old' overflow bucket
                    self.table[ht_index].slots[length][key] = Slot()
                    self.table[ht_index].slots[length][key].key = data
                    self.table[ht_index].slots[length][key].value = value
                    # self.table[ht_index].slots[key][length] = data
                    self.table[ht_index].count += 1
                    self.SetOverflowPointer(ht_index, key)
                else:
                    # new overflow bucket
                    self.table[ht_index].slots[0][key] = Slot()
                    self.table[ht_index].slots[0][key].key = data
                    self.table[ht_index].slots[0][key].value = value
                    self.table[ht_index].count += 1
                    self.SetOverflowPointer(ht_index, key)
            else:
                # overflow is false, collission still true
                length = self.table[ht_index].count
                # print length
                # print ht_index, slot_index
                self.table[ht_index].slots[length][key] = Slot()
                self.table[ht_index].slots[length][key].key = data
                self.table[ht_index].slots[length][key].value = value
                self.table[ht_index].count += 1
                # print item.child()
                # print item.count
        else:
            #brand new key
            ht_index, slot_index = self.CheckAvailableBuckets(key, 0, TableSize)
            self.table[ht_index].slots[slot_index][key] = Slot()
            self.table[ht_index].slots[slot_index][key].key = data
            self.table[ht_index].slots[slot_index][key].value = value
            self.table[ht_index].count += 1

    def SetOverflowPointer(self, index, key):
        ht_index, temp = self.CheckAvailableBuckets(key,
            0, TableSize)
        self.table[index].overflow = ht_index
        self.table[ht_index].overflow = index

    def CollissionCheck(self, key):
        # may not be a slot yet
        for pos, item in enumerate(self.table):
            for ps, ite in enumerate(item.slots):
                if key in item.slots[ps]:
                    return ite, pos, ps
                    # print item.slots[ps].keys()
                    # return item.slots[ps].key, ps, pos
                    #return item.slots[ps][key].key
                #return item.slots.key, pos
        return None

    def CheckAvailableBuckets(self, key, minimum, maximum):
        # self.table[5].slots[1][key] = Slot()
        # self.table[5].slots[1][key].key = key
        for i in range(minimum, maximum):
            # print minimum, maximum
            for j in range(0, 3):
                # print self.table[i].slots[j].keys()
                # print len(self.table[i].slots[0])

                if key in self.table[i].slots[j].keys():
                    return i, j
                if len(self.table[i].slots[0]) is 0:
                    return i, j
                #     return j
        # # Should never get here
        raise Exception('Ran out of overflow buckets!')


ht = HashTable()

# s = Slot()
# s[0]['key'] = Slot()
# ht.Insert('key', 'value', 'keyvalue')
ht.DataIn()

ht.GenerateReport()
ht.Search()

# ht.SaveReport()
# print ht.table
# ht.RestoreReport()
# # print ht.table
# print ht.table[0].slots[0][14].key
# fuck = ht.CollissionCheck('key')
# ht.CheckAvailableBuckets('key', 0, 20)
# ht.Insert('key','value','keyvalue')
# ht.Insert('key','value','keyvalue')
# ht.Insert('key','value','keyvalue')
# ht.Insert('key','value','keyvalue')
# ht.Insert('key','value','keyvalue')
# ht.Insert('key','value','keyvalue')
# ht.Insert('key','value','keyvalue')
# ht.Insert('key','value','keyvalue')
# ht.Insert('key','value','keyvalue')
# print ht.table[0].slots[0]['key'].value
# print ht.table[0].slots[1]['key'].value
# print ht.table[0].slots[2]['key'].value
# print ht.table[19].slots[0]['key'].value
# print ht.table[19].slots[1]['key'].value
# print ht.table[19].slots[2]['key'].value
# print ht.table[20].slots[0]['key'].value
# print ht.table[19].slots[1]['key'].value
# s['key'] = 'v'
# print s
#ht.CollissionCheck('k')
# ht.Insert('k', 'k', 'kv')
# ht.Insert('j', 's-', 'kv')
# print ht.table[0].slots[0].key
#print data
# ht.DataIn()

# ht.GenerateReport()

# ht.SaveReport()

# if ht.table is not None:
#     raise Exception('The hash table must be deleted from memory.')

# ht.RestoreReport()

# print('{0:20} {1}'.format(' ','hash table'.center))

File: assign1/cs/hashtable.cs
/*
@author Adam Voliva
@description CISP 430 Data Structures
@name Assignment 1 Hash Table
@date February 6, 2013
@file hashtable.cs
*/


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters;
using System.Runtime.Serialization.Formatters.Binary;

/*
const int MaxSlots = 3;
const int MaxBuckets = 30;
const int TableSize = 20;
const int OverflowSize = 10;
 * */

namespace hashtable
{
    [Serializable]
    class Slot
    {
        public string key;
        public string value;
        public Slot()
        {
            this.key = null;
            this.value = null;
        }
    }

    [Serializable] 
    class Bucket
    {
        public int count;
        public int overflow;
        public List<Dictionary<string, Slot>> slots;
        //public Dictionary<Slot, string> somedict;
        public Bucket()
        {
            this.count = 0;
            this.overflow = -1;
            this.slots = new List<Dictionary<string, Slot>>();
            for (int i = 0; i < 3; ++i)
            {
                this.slots.Add(new Dictionary<string, Slot>());
            }

        }
    }

    [Serializable] 
    class HashTable
    {
        public List<Bucket> table;
        public HashTable()
        {
            this.table = new List<Bucket>();
            for (int i = 0; i < 30; ++i)
            {
                this.table.Add(new Bucket());
            }
            //List<Bucket> table = new List<Bucket>();
        }

        public void SaveReport()
        {
            Stream stream = null;
            try
            {
                IFormatter formatter = new BinaryFormatter();
                stream = new FileStream(@"C:\Users\Owner\Documents\cisp430\git\assign1\hashtable.txt", FileMode.Create, FileAccess.Write, FileShare.None);
                formatter.Serialize(stream, this.table);
            }
            catch
            {
                // do nothing
            }
            finally
            {
                if (null != stream)
                    stream.Close();
            }
        }
        public void GenerateReport()
        {
            List<String> lines = new List<String>();
            //lines.Add("Hash Table"
            lines.Add(String.Format("{0, 45}", "Hash Table"));
            lines.Add(String.Format("{0, 49}", "Verification Report"));
            lines.Add(String.Format("{0, 46}", "Before Report"));
            int i = 0;
            foreach (Bucket item in this.table)
            {
                lines.Add(String.Format("Bucket {0}", i + 1));
                int j = 0;
                foreach (Dictionary<string, Slot> dict in item.slots)
                {
                    int k = 0;
                    foreach (KeyValuePair<string, Slot> kv in item.slots[j])
                    {
                        
                        if (item.slots[j][kv.Key].key != null)
                        {
                            lines.Add(String.Format("\tSlot {0}: {1}{2}",
                                j + 1,
                                item.slots[j][kv.Key].key,
                                item.slots[j][kv.Key].value
                            ));
                        }
                        ++k;
                        //Console.WriteLine(k);
                     
                    }
                    if (item.slots[j].Count == 0)
                    {
                        lines.Add(String.Format("\tSlot {0}: None", j + 1));
                    }
                   // Console.WriteLine(j);
                    ++j;
                }
                if (this.table[i].overflow != -1)
                {
                    lines.Add(String.Format("\tOverflow Pointer: {0}", this.table[i].overflow + 1));
                }
                else
                {
                    lines.Add(String.Format("\tOveflow Pointer: {0}", "None"));
                }
                lines.Add(String.Format("{0}", "\n"));
                ++i;
            }
            string[] text = lines.ToArray();
            System.IO.File.WriteAllLines(@"C:\Users\Owner\Documents\cisp430\git\assign1\before.txt", text);
        }
          
        public void DataIn()
        {
            //Object file = System.IO.File.Open('datain.dat', 'r+');
            //FileStream file = new FileStream(@"c:\datain.dat", FileMode.Open, FileAccess.Read);
            string[] lines = System.IO.File.ReadAllLines(@"C:\Users\Owner\Documents\cisp430\git\assign1\datain.dat");
            foreach (string line in lines)
            {
                this.Insert(this.Hash(line.Substring(0, 10)), line.Substring(10, 20), line.Substring(0, 10));
            }
        }

        public string Hash(string key)
        {
            return (((int)key[2] + (int)key[4] + (int)key[6]) % 20).ToString();
        }

        public void Insert(string key, string value, string data)
        {
            Tuple<Object, int> tup = this.CollissionCheck(key);
            if (tup.Item1 != null)
            {
                //collission is true
                //Dictionary<string, string> item = (Dictionary<string, string>)tup.Item1;
                int index = tup.Item2;
                if (this.table[index].count > 2)
                {
                    //overflow is true
                    //Console.WriteLine("overflow is true");
                    index = this.CheckAvailableBuckets(key, 20, 30);
                    int length = this.table[index].count;
                    Console.WriteLine("length is " + length);
                    if (length == 3)
                    {
                        //overflow of overflow is true
                        //Console.WriteLine("overflow of overflow is true");
                        index = this.CheckAvailableBuckets(key, index + 1, 30);
                        length = this.table[index].count;
                    }
                    if (length > 0)
                    {
                        //insert into old overflow bucket
                        //Console.WriteLine("insert into old overflow bucket");
                        this.table[index].slots[length][key] = new Slot();
                        this.table[index].slots[length][key].key = key;
                        this.table[index].slots[length][key].value = value;
                        this.table[index].count += 1;
                    }
                    else
                    {
                        // new overflow bucket
                        //Console.WriteLine("new overflow bucket");
                        this.table[index].slots[0][key] = new Slot();
                        this.table[index].slots[0][key].key = key;
                        this.table[index].slots[0][key].value = value;
                        this.table[index].count += 1;
                        this.SetOverflowPointer(index, key);
                    }
                }
                else
                {
                    // overflow is false, collission still true
                    //Console.WriteLine("overflow is false, collission still true");
                    int length = this.table[index].count;
                    //Console.WriteLine(length);
                    //Console.WriteLine(key);
                    this.table[index].slots[length][key] = new Slot();
                    this.table[index].slots[length][key].key = key;
                    this.table[index].slots[length][key].value = value;
                    this.table[index].count += 1;
                    //Console.WriteLine("index is " + index);
                }
            }
            else
            {
                // brand new key
                //Console.WriteLine("brand new key");
                int index = this.CheckAvailableBuckets(key, 0, 20);
                int length = this.table[index].count;
                this.table[index].slots[length][key] = new Slot();
                this.table[index].slots[length][key].key = key;
                this.table[index].slots[length][key].value = value;
                this.table[index].count += 1;
                //Console.WriteLine("index is " + index);
            }
        }

        public void SetOverflowPointer(int index, string key)
        {
            int of_index = this.CheckAvailableBuckets(key, 0, 20);
            this.table[index].overflow = (this.table[index].overflow == -1) ? of_index : this.table[index].overflow;
            this.table[of_index].overflow = (this.table[of_index].overflow == -1) ? index : this.table[of_index].overflow;
            // this is for the pointer of overflow of overflows 
            of_index = this.CheckAvailableBuckets(key, 20, 30);
            if (of_index != index)
                this.table[index].overflow = of_index;
            Console.WriteLine(of_index);
            Console.WriteLine("index = " + index);
        }

        public Tuple<Object, int> CollissionCheck(string key)
        {
            int i = 0;
            foreach(Bucket item in this.table)
            {
                foreach(Dictionary<string, Slot> dict in item.slots)
                {
                    //foreach(string k in dict.Keys)
                    //{
                        if (dict.ContainsKey(key))
                        {
                            return new Tuple<Object, int>(dict, i);
                        }
                    //}
                }
                ++i;
            }
            return new Tuple<Object, int>(null, 0);
        }
        
        public int CheckAvailableBuckets(string key, int min, int max)
        {
            for (int i = min; i < max; ++i)
            {
                for (int j = 0; j < 3; ++j) 
                {
                    //foreach(Slot k in this.table[i].slots[j].Keys)
                    //{
                        if (this.table[i].slots[j].ContainsKey(key))
                        {
                            //Console.WriteLine("now i is " + i + " and j is " + j);
                            return i;
                        }
                        if (this.table[i].slots[0].Count == 0)
                        {
                            //Console.WriteLine("i is " + i + ", j is " + j);
                            return i;
                        }

                    //}
                }
            }
            return 0;
        }
    }

    class TestClass
    {
        static void Main(string[] args)
        {
            // Display the number of command line arguments:
            //Bucket shit = new Bucket();
            /*foreach(Slot key in shit.somedict.Keys)
            {
                Console.WriteLine(key.key);
            }*/

            HashTable ht = new HashTable();
            /*
            ht.Insert("key", "data", "keydata");
            ht.Insert("key", "data", "keydata");
            ht.Insert("key", "data", "keydata");
            ht.Insert("key", "data", "keydata");
            ht.Insert("key", "data", "keydata");
            ht.Insert("key", "data", "keydata");
            ht.Insert("key", "data", "keydata");
             * */

            ht.DataIn();
            ht.GenerateReport();
            ht.SaveReport();

            Console.ReadLine();
        }
    }
}